src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp

```diff
@@ -687,63 +698,145 @@ size_t ObjectStorageQueueMetadata::unregisterActive(const StorageID & storage_id
-                found = true;
+            if (remove_metadata_if_no_registered && count == 0)
+            {
+                LOG_TRACE(log, "Removing all metadata in keeper by path: {}", zookeeper_path.string());
+                if (supports_remove_recursive)
+                {
+                    requests.push_back(zkutil::makeCheckRequest(registry_path, stat.version));
+                    requests.push_back(zkutil::makeRemoveRecursiveRequest(zookeeper_path, -1));
+                }
+                else
+                {
+                    requests.push_back(zkutil::makeCheckRequest(registry_path, stat.version));
+                    requests.push_back(zkutil::makeCreateRequest(drop_lock_path, "", zkutil::CreateMode::Ephemeral));
+                }
+                code = zk_client->tryMulti(requests, responses);
+            }
             else
             {
-                if (!new_registry_str.empty())
-                    new_registry_str += ",";
-                new_registry_str += elem;
-                count += 1;
+                code = zk_client->trySet(registry_path, new_registry_str, stat.version);
             }
         }
-        if (!found)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot unregister: table '{}' is not registered", self.table_id);
-
-        code = zk_client->trySet(registry_path, new_registry_str, stat.version);
+        catch (const zkutil::KeeperMultiException & e)
+        {
+            if (Coordination::isHardwareError(e.code))
+            {
+                LOG_TEST(log, "Lost connection to zookeeper, will retry");
+                continue;
+            }
+            throw;
+        }
+        catch (const zkutil::KeeperException & e)
+        {
+            if (Coordination::isHardwareError(e.code))
+            {
+                LOG_TEST(log, "Lost connection to zookeeper, will retry");
+                continue;
+            }
+            throw;
+        }
 
         if (code == Coordination::Error::ZOK)
         {
             LOG_TRACE(log, "Table '{}' has been removed from the registry", self.table_id);
+
+            if (!supports_remove_recursive && remove_metadata_if_no_registered && count == 0)
+            {
+                /// Take a drop lock and do recursive remove as a separate request.
+                /// In case of unsupported "remove_recursive" feature, it will
+                /// do getChildren and remove them one by one.
+                auto drop_lock = zkutil::EphemeralNodeHolder::existing(drop_lock_path, *zk_client);
+                try
+                {
+                    zk_client->removeRecursive(zookeeper_path);
+                }
+                catch (const zkutil::KeeperMultiException & e)
+                {
+                    if (Coordination::isHardwareError(e.code))
+                    {
+                        LOG_TEST(log, "Lost connection to zookeeper, will retry");
+                        continue;
```

Comment:
Is the idea that the `drop_lock_path` node is being removed in the `EphemeralNodeHolder()` destructor when `drop_lock` goes out of scope? I don't usually work with Zookeeper, so I'm asking.
If an exception occurs during `zk_client->removeRecursive(zookeeper_path);`, and the loop is retried, do we still want to remove the ephemeral node in the error path? There is `setAlreadyRemoved()` method that makes it no-op.